### PR TITLE
feat: ✨ Add Stalker Component

### LINKED ISSUE #34: Allow Catniss to "stalk" other players (if a match isn't found)
If there is an odd number of players and Catniss cannot find a match that round, she should be able to select a matching pair to view their entire chat log during the chatting phase. so that she doesn't feel left out that round.

**Tasks:**

- [x] Assign random player as stalker after checking odd # of players in Backend
- [x] Display chats as buttons with who is talking with who in Frontend
- [x] Chat buttons should be clickable which will take user to selected chat in Backend
- [x] Stalker should be taken to stalker component in Backend.
- [x] Stalker should join a chat anonymously with read-only function Backend.

**Acceptance Criteria**

- If there is an even number of players in the lobby, this pseudo-role does not exist until there is an odd number.
- If there is an odd number of players, one player selected at random will become a stalker
- If the user is the stalker, they should be shown the current pairs in the round to which they can select which pair they want to stalk
- When stalking, they should be able to see the current chatting between the selected pair
- If the user is not the stalker, they would play the chatting phase normally
- If the user is being stalked, they should not be able to see that they are being stalked until the voting phase
- A hacker shouldn't be able to see if they are being stalked
- A hacker who is the stalker should not be able to stock more than one pair
- A hacker w should not be able to influence who is the stalker

### PR DESCRIPTION
**Pull request by njumaah**
Adds stalker component to be used for lobby.

### CODE REVIEW

**Comment by nstringham** [comment]
related to #34

**Review by nstringham — CHANGES_REQUESTED** [review]
please pull from development

File: client/src/components/ChatRoom.svelte [hunk 1]
```diff
@@ -84,28 +104,50 @@
 </script>
 
 <div class="chatroom">
-  {#if countdown !== undefined}
+  {#if isStalker}
+    {#if countdown !== undefined}
+      <p class="countdown">{countdown}</p>
+    {/if}
```
Comment (gchamb) [inline client/src/components/ChatRoom.svelte]: You can render this once since it doesn't matter if it is the stalker view or chatroom view.

File: client/src/components/ChatRoom.svelte [hunk 2]
```diff
@@ -84,28 +104,50 @@
...
+          <p class="partner-name">{firstPairInfo.displayName}</p>
+          <p class="partner-msg">{message.text}</p>
+        {:else}
+          <p class="partner-name">{secondPairInfo.displayName}</p>
+          <p class="partner2-msg">{message.text}</p>
+        {/if}
+      {/each}
+    </div>
+  {:else}
     <p class="countdown">{countdown}</p>
```
Comment (gchamb) [inline client/src/components/ChatRoom.svelte]: You can render this once since it doesn't matter if it is the stalker view or chatroom view.

File: client/src/components/Stalker.svelte [hunk 3]
```diff
@@ -0,0 +1,106 @@
+<script lang="ts">
+  import type { ChatRoom as ChatRoomType, ChatMessage, Lobby } from "$lib/firebase/firestore-types/lobby";
+  import { stalkChatroom } from "$lib/firebase/firestore-functions";
+  import { getChatRoomCollection, getChatRoomMessagesCollection } from "$lib/firebase/firestore-collections";
+  import { onSnapshot, query, orderBy, collection } from "firebase/firestore";
+  import { onMount, onDestroy } from "svelte";
+  import type { Unsubscribe } from "firebase/auth";
+  // import { auth } from "$lib/firebase/app";
```
Comment (gchamb) [inline client/src/components/Stalker.svelte]: unused

File: client/src/components/Stalker.svelte [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  export let lobbyCode: string;
+
+  let unsubscribeChatrooms: Unsubscribe | undefined = undefined;
+  let chatrooms: ChatRoomType[] = [];
+  let chatroomsIds: string[] = [];
+  let chatMessages: ChatMessage[] = [];
+
+  onMount(async () => {
+    const chatCollection = getChatRoomCollection(lobbyCode);
+    unsubscribeChatrooms = onSnapshot(chatCollection, (chatSnapshot) => {
```
Comment (gchamb) [inline client/src/components/Stalker.svelte]: do we need to be subscribed to the chatrooms collection? Will there be any changes we need to listen for?

File: client/src/components/Stalker.svelte [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        return chatDoc.data().viewers.includes($user!.uid);
+      });
+      if (room !== undefined) {
+        onSnapshot(
+          query(getChatRoomMessagesCollection(lobbyCode, room.id), orderBy("timestamp", "asc")),
+          (collection) => {
+            chatMessages = collection.docs.map((d) => d.data());
+          }
+        );
+      }
```
Comment (gchamb) [inline client/src/components/Stalker.svelte]: why are you subscribing to the chat messages if you're using them only to check it's length?

File: client/src/components/Stalker.svelte [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    {#each chatrooms as chatroom, i}
+      <button
+        on:click={() => {
+          onClickChat(chatroomsIds[i]);
+        }}
+        class="chatRoom"
+      >
+        <span class="pair">{findDisplayName(chatroom.pair[0]) + " & " + findDisplayName(chatroom.pair[1])}</span>
+      </button><br />
+    {/each}
```
Comment (gchamb) [inline client/src/components/Stalker.svelte]: Since you're displaying the chatrooms with their associated pair shouldnt you check if chatrooms.length != 0 instead of `line 61`

File: client/src/lib/firebase/firestore-functions.ts [hunk 7]
```diff
@@ -1,9 +1,12 @@
 import { functions } from "$lib/firebase/app";
 import { httpsCallable } from "firebase/functions";
-import type { LobbyRequest } from "./firestore-functions-types";
+import type { LobbyRequest, ChatRequest, StalkChatroomRequest } from "./firestore-functions-types";
 import type { ChangeAvatarData } from "./functions-types/avatar";
 
 export const startGame = httpsCallable<LobbyRequest, void>(functions, "startGame");
 export const joinLobby = httpsCallable<LobbyRequest, void>(functions, "joinLobby");
+export const stalkChatroom = httpsCallable<StalkChatroomRequest, void>(functions, "stalkChatroom");
+export const deleteChatRooms = httpsCallable<LobbyRequest, void>(functions, "deleteChatRooms");
```
Comment (gchamb) [inline client/src/lib/firebase/firestore-functions.ts]: this shouldn't be here. Are you merged with development?

File: client/src/routes/game/+page.svelte [hunk 8]
```diff
@@ -1,6 +1,8 @@
 <script lang="ts">
   import Prompt from "$components/Prompt.svelte";
   import LobbyComponent from "$components/Lobby.svelte";
+  import Stalker from "$components/Stalker.svelte";
+
```
Comment (gchamb) [inline client/src/routes/game/+page.svelte]: remove space

File: functions/src/chat.ts [hunk 9]
```diff
@@ -31,10 +35,14 @@ export async function deleteChatRooms(lobbyData: Lobby, lobbyDoc: DocumentRefere
         players[playerIndex].promptAnswer = value;
       }
 
+      // delete stalker
+      const stalkers = await getPrivatePlayerCollection(lobbyDoc).where("stalker", "==", true).get();
```
Comment (gchamb) [inline functions/src/chat.ts]: put it in transaction

**Comment by gchamb** [comment]
in firestore-functions-types.ts you have the type generatedpairs still in there please remove

File: functions/src/lobby.ts [hunk 10]
```diff
@@ -188,8 +188,12 @@ export const collectPromptAnswers = functions.firestore
 
       transaction.update(lobbyDocRef, { state: "CHAT" });
 
-      // TODO: check if we have a stalker
-      const { pairs } = generatePairs(lobbyData);
+      const { pairs, stalker } = generatePairs(lobbyData);
+      functions.logger.info(stalker);
```
Comment (gchamb) [inline functions/src/lobby.ts]: you can remove this

File: functions/src/stalker.ts [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+    throw new functions.https.HttpsError("permission-denied", "User is not Authenticated");
+  }
+
+  if (!isStalkChatroomRequest(data)) {
+    throw new functions.https.HttpsError("invalid-argument", "Data is not of StalkChatroomRequest type");
+  }
+
+  const lobbyDocRef = lobbyCollection.doc(data.code);
+  return db.runTransaction(async (transaction) => {
+    const chatRooms = getChatRoomCollection(lobbyDocRef);
```
Comment (gchamb) [inline functions/src/stalker.ts:18]: use transaction

File: functions/src/stalker.ts [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+    const chatRoomsRef = chatRooms.doc(data.chatId);
+    const chatRoomsSnapshot = await transaction.get(chatRooms.where("viewers", "array-contains", auth.uid));
+    if (!chatRoomsSnapshot.empty) {
+      throw new functions.https.HttpsError("already-exists", "User already stalking a chat");
+    }
+    const chatRoom = await transaction.get(chatRooms.doc(data.chatId));
```
Comment (gchamb) [inline functions/src/stalker.ts]: use the variable you made on `line 19`

`

File: firestore.rules [hunk 13]
```diff
@@ -34,12 +34,17 @@ service cloud.firestore {
         function inPair(){
           return request.auth.uid in roomData().pair 
         }
+        function isStalker(){
+          return get(/databases/$(database)/documents/lobbies/$(code)/privatePlayers/$(request.auth.uid)).data.stalker == true;
+        }
```
Comment (gchamb) [inline firestore.rules:39]: why is this needed?

Comment (gchamb) [inline firestore.rules:37]: why is this needed?

File: firestore.rules [hunk 14]
```diff
@@ -34,12 +34,17 @@ service cloud.firestore {
...
         function canRead(){
           return inPair() || request.auth.uid in roomData().viewers ;
         }
+
+
         // firestore didn't like me using the above function for chatroom
         // I think its because the info is in the resource object
-        allow read: if request.auth.uid in resource.data.pair || request.auth.uid in resource.data.viewers ;
+        allow read: if request.auth.uid in resource.data.pair || request.auth.uid in resource.data.viewers || isStalker() ;
```
Comment (gchamb) [inline firestore.rules:45]: `request.auth.uid in resource.data.viewers` already allow stalker to read chat. It wouldn't make it to your isStalker function because it'd already be true.

File: client/src/lib/firebase/firestore-functions.ts (hunk 7 above)
Comment (njumaah) [reply client/src/lib/firebase/firestore-functions.ts]: i am merged with dev, where should it be?

Comment (gchamb) [reply client/src/lib/firebase/firestore-functions.ts]: Remove line 9

File: firestore.rules (hunk 14 above)
Comment (njumaah) [reply firestore.rules:45]: image
We haven't added Stalker to the viewers list at this point yet! We need the stalker to be able to see the list of all chats _before_ we add him to a viewer list of a specific chat. See image for what happens if we don't have this.

File: firestore.rules (hunk 13 above)
Comment (njumaah) [reply firestore.rules:37]: See reason in resolved review below

File: functions/src/firebase-functions-types.ts [hunk 15]
```diff
@@ -6,3 +6,37 @@ export function isLobbyRequest(data: unknown): data is LobbyRequest {
...
+    data != null &&
+    typeof data === "object" &&
+    "code" in data &&
+    typeof (data as ChatRequest).code === "string" &&
+    "roomId" in data &&
+    typeof (data as ChatRequest).roomId === "string" &&
+    "message" in data &&
+    typeof (data as ChatRequest).message === "string"
+  );
+}
```
Comment (nstringham) [inline functions/src/firebase-functions-types.ts:37]: this is not used

File: functions/src/firebase-functions-types.ts [hunk 16]
```diff
@@ -6,3 +6,37 @@ export function isLobbyRequest(data: unknown): data is LobbyRequest {
...
+
+export type GeneratedPairs = {
+  pairs: { one: string; two: string }[];
+  stalker?: string;
+};
```
Comment (nstringham) [inline functions/src/firebase-functions-types.ts:42]: this got moved to util.ts

File: functions/src/stalker.ts [hunk 17]
```diff
@@ -0,0 +1,32 @@
+import * as functions from "firebase-functions";
+import { db } from "./app";
+import { isStalkChatroomRequest } from "./firebase-functions-types";
+import { getChatRoomCollection, lobbyCollection } from "./firestore-collections";
+
+export const stalkChatroom = functions.https.onCall((data: unknown, context) => {
```
Comment (nstringham) [inline functions/src/stalker.ts:6]: please specify a return type like in 191b0bf80c59858ed7b9e1b52d2e04ce9b9f2f9f

File: client/src/lib/firebase/firebase-functions.ts [hunk 18]
```diff
@@ -1,10 +1,17 @@
...
+  ChatRequest,
+  StalkChatroomRequest,
+} from "./firebase-functions-types";
 import type { ChangeAvatarData } from "./functions-types/avatar";
 
 export const createLobby = httpsCallable<void, LobbyCreationResponse>(functions, "createLobby");
 export const startGame = httpsCallable<LobbyRequest, void>(functions, "startGame");
 export const joinLobby = httpsCallable<LobbyRequest, void>(functions, "joinLobby");
+export const stalkChatroom = httpsCallable<StalkChatroomRequest, void>(functions, "stalkChatroom");
+export const addChatMessage = httpsCallable<ChatRequest, void>(functions, "addChatMessage");
```
Comment (nstringham) [inline client/src/lib/firebase/firebase-functions.ts:15]: addChatMessage is not used and should be deleted

File: client/src/components/Stalker.svelte [hunk 19]
```diff
@@ -0,0 +1,91 @@
+<script lang="ts">
+  import type { ChatRoom as ChatRoomType, Lobby } from "$lib/firebase/firestore-types/lobby";
+  import { stalkChatroom } from "$lib/firebase/firebase-functions";
+  import { getChatRoomCollection } from "$lib/firebase/firestore-collections";
+  import { getDocs, onSnapshot, query, where } from "firebase/firestore";
+  import { onMount } from "svelte";
+  import ChatRoom from "./ChatRoom.svelte";
```
Comment (nstringham) [inline client/src/components/Stalker.svelte:7]: lets put the component imports at the top

File: client/src/components/Stalker.svelte [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+  let chatroomsIds: string[] = [];
+  let selectedChat: string | undefined = undefined;
+
+  onMount(async () => {
+    const chatCollection = getChatRoomCollection(lobbyCode);
+    const chatSnapshot = await getDocs(chatCollection);
+    chatroomsIds = chatSnapshot.docs.map((room) => room.id);
+    chatrooms = chatSnapshot.docs.map((room) => room.data());
+
+    onSnapshot(query(chatCollection, where("viewers", "array-contains", $user!.uid)), (snapshot) => {
```
Comment (nstringham) [inline client/src/components/Stalker.svelte:23]: see if you can use something more robust instead of asserting that it is not null. maybe an if statment that throws an exception?

File: client/src/components/Stalker.svelte [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+   * takes uid of single user, then uses the index of uid to find and return display name
+   */
+  function findDisplayName(uid: string): string {
+    return lobby.players[lobby.uids.indexOf(uid)].displayName;
+  }
+
+  // takes chatid to send a stalk chatroom request
+  function onClickChat(chatId: string) {
+    const stalkChatroomRequest = { code: lobbyCode, chatId };
+    return stalkChatroom(stalkChatroomRequest);
```
Comment (nstringham) [inline client/src/components/Stalker.svelte:40]: we should be checking for errors here and displaying them to the user

File: client/src/components/Stalker.svelte [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+  }
+
+  // onsnapshot subscribe to list of all chatrooms where you a viewer, if not empty switch to show chatroom
```
Comment (nstringham) [inline client/src/components/Stalker.svelte:43]: is this comment still correct?

**Review by nstringham — CHANGES_REQUESTED** [review]

File: client/src/lib/firebase/firebase-functions.ts (hunk 18 above)
Comment (nstringham) [reply client/src/lib/firebase/firebase-functions.ts:15]: 154d5f49ad7b68024548917c81138dba3ae83e3b

File: functions/src/firebase-functions-types.ts (hunk 16 above)
Comment (nstringham) [reply functions/src/firebase-functions-types.ts:42]: e4a8173eb10cef77ce3772d92983b8d0e534717b

File: functions/src/firebase-functions-types.ts (hunk 15 above)
Comment (nstringham) [reply functions/src/firebase-functions-types.ts:37]: e4a8173eb10cef77ce3772d92983b8d0e534717b

File: functions/src/stalker.ts (hunk 17 above)
Comment (nstringham) [reply functions/src/stalker.ts:6]: 33ab56c73177cef7ca749a4d9d4b5adebb94d63e

File: client/src/components/Stalker.svelte (hunk 22 above)
Comment (nstringham) [reply client/src/components/Stalker.svelte:43]: c05ddcf263383dc29c6380e525cf184b297990f8

**Comment by mrjoshua520** [comment]
close in favor of #112
